src/main/java/StringManipulation.java

```diff
@@ -1,33 +1,133 @@
+
+        // split the string by whitespace(s), period, comma, and other symbols
+        // and stores each word into a String array
+        String[] words = stringToManipulate.split("[\\s.(),:?!]+");
+
+        // returns the length of array as the word count
+        return words.length;
     }
 
     @Override
     public String removeNthCharacter(int n, boolean maintainSpacing) {
-        return null;
+
+        // If the current stringToManipulate is null, throws an exception
+        if(stringToManipulate == null){
+            throw new NullPointerException("The string is null.");
+        }
+        // If the given n is smaller or equal to 0, throws IllegalArgumentException
+        if(n <= 0){
+            throw new IllegalArgumentException("n must be greater than zero.");
+        }
+        // If the given n is larger than the length of the string, throws IndexOutOfBoundsException
+        if(n > stringToManipulate.length()){
+            throw new IndexOutOfBoundsException("n must be less than or equal to the length of the string.");
+        }
+
+        StringBuilder str = new StringBuilder();
+        int length = stringToManipulate.length();
+
+        // Traverse the entire stringToManipulate
+        for(int i = 1; i <= length ;i++){
+            // As long as the index of stringToManipulate is NOT divisible by n, append the
+            // corresponding character to the str
+            if(i % n != 0) {
+                str.append(stringToManipulate.charAt(i - 1));
+            }
+            // if the index is divisible by n AND maintainSpacing is true, append a space
+            else if(maintainSpacing){
+                str.append(' ');
+            }
+        }
+
+        return str.toString();
     }
 
     @Override
-    public String[] getSubStrings(int startWord, int endWord) {
-        return null;
+    public String[] getSubStrings(int startWord, int endWord){
+
+        // If the current stringToManipulate is null, throws an exception
+        if(stringToManipulate == null){
+            throw new NullPointerException("The string is null.");
+        }
+        // If the startWord or the endWord is zero or negative, OR the endWord is smaller than the startWord,
+        // throws an exception
+        if(startWord <= 0 || endWord <= 0 || startWord > endWord)
+        {
+            throw new IllegalArgumentException("Invalid argument(s) were entered.");
+        }
+
+        // Gets the word count by using split()
+        String[] words = stringToManipulate.split("[\\s.,_:?!]+");
+
+        // If the endWord is larger than the length of string, throws an exception
+        if(endWord > words.length)
+        {
+            throw new IndexOutOfBoundsException("The endWord entered is out of bounds");
+        }
+
+        // copies the words in the specified range into a new array called substring
+        String[] substring = Arrays.copyOfRange(words, startWord-1, endWord);
+
+        return substring;
     }
```

Comment:
Great job on the getSubStrings method!

Appreciate how you've implemented the necessary validations and error handling. The code is clear and easy to understand, thanks to the informative exception messages you've provided. It helps users quickly identify and address any issues with their input.

Additionally, your use of the split() method to extract words from the string is a smart approach. It effectively handles various whitespace and punctuation characters, ensuring accurate word extraction.

_Refactoring Opportunity_:

One small refactoring opportunity I noticed is in the if condition for validating startWord and endWord. Instead of checking for startWord <= 0 || endWord <= 0, you could simplify it by using startWord < 1 || endWord < 1. This modification aligns with the subsequent comment explaining that zero values are invalid, and it enhances readability by removing unnecessary repetition.

Reply:
Thank you for your compliment and good suggestion! I really appreciate it:)

While I agree with you that (startWord < 1 || endWord < 1) looks simpler and might improve readability, I would go with my original code (startWord <= 0 || endWord <= 0) this time.

The reason is that it aligns with the requirement description. The requirement description for the method uses the following expressions: 
     * throws IllegalArgumentException
     *             If either "startWord" or "endWord" are invalid (i.e.,
     *             "startWord" <= 0, "endWord" <= 0, or "startWord"
     *             > "endWord")
Because of these expressions, I feel using the same expressions in the code would make it easier for anyone to check later if we have implemented the exception handling correctly.

Having said that, I know the requirement description would not be always clear like this and your suggestion definitely opened my eyes to other ways that could be used and possibly better in a lot of situations:)